File: #20_Valid_Parentheses/Solution.java
import java.util.*;

public class Solution {
    public boolean isValid(String s) {
    	Deque<Character> stk = new LinkedList<Character>();
    	for (int i = 0; i < s.length(); i++) {
    		char c = s.charAt(i);
    		if (c == '(' || c == '{' || c == '[') stk.push(c);
    		else {
    			if (stk.isEmpty()) return false;
    			char p = stk.pop();
    			if (p != '{' && c == '}') return false;
    			if (p != '(' && c == ')') return false;
    			if (p != '[' && c == ']') return false;
    		}
    	}
    	return stk.isEmpty();
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	String s = "{[}]";
    	System.out.println(sol.isValid(s));
    }
}

File: #20_Valid_Parentheses/Solution.py
class Solution:
	# @param {string} s
	# @return {boolean}
	def isValid(self, s):
		if not s:
			return True
		stk = []
		dict = {')': '(', '}': '{', ']': '['}
		for c in s:
			if c == '(' or c == '{' or c == '[':
				stk.append(c)
			else:
				if not stk or stk.pop() != dict[c]:
					return False
		return not stk

sol = Solution()
print sol.isValid("(")